### PR TITLE
recognize Float as rational

### LINKED ISSUE #23731: `is_zero` broken for integer/float comparison
Not sure if this is a recent regression, but comparing an integer-assumption symbol to a Float (or a non-integer to an `Integer`) returns the wrong value:

```python
>>> i = Symbol('x', integer=True)
>>> Eq(i, 1.0)
False
>>> Eq(i, 1)
Eq(x, 1)

>>> ni = Symbol('x', integer=False)
>>> Eq(ni, 1)
False
>>> Eq(ni, 1.0)
Eq(x, 1.0)
```

Upon further inspection, it has to do with the difference being checked in `is_eq`: `_dif.is_zero` returns False, even though that should not always be true.

### PR DESCRIPTION
**Pull request by smichr**
#### References to other Issues or PRs

alternative to #25856 and #25864
fixes #23731 

#### Brief description of what is fixed or changed

That Float may be an approximation of an irrational number (or repeating binary fraction) does not make it any less a Rational number (though it is much more granular than the Rational representation used in SymPy). A Float is a computationally efficient representation of a number in a rational form.

The attributes have been updated accordingly.
```python
>>> Float(1).is_odd == Float(1).is_rational == Float(1).is_integer == True
True
>>> Float(3).is_prime
True
```

There was a long discussion of whether `Float` should be classified as `rational` [here](https://groups.google.com/g/sympy/c/RUYOxxwMgac/m/4-1JXtny32EJ). One major concern was that float/Float objects do not behave as Rationals in operations, e.g.
```python
>>> Float(1/3)+Float(1/10)-Float(1/10)-Float(1/3)
5.55111512312578e-17
>>> _.is_rational
True
>>> Float(1/3)+Float(1/10) - (Float(1/10)+Float(1/3))
0
>>> _.is_rational
True
```
Allowing a object to be classified as rational, however, does not imply how it behaves in computation *in particular* only in *general*. In general `rational +/- rational` is `rational`. Anyone using SymPy would be expected to write code to take the appropriate action for a given value type. The user is also expected to understand the difference between an exact result derived from a Float-free expression and a result from an expression containing Floats. The user is also expected to understand that associative properties do not apply to Float-rationals like they do to Rational-rationals (as shown above).

* It was the assumption system that was running into problems when being used to evaluate expressions like `Eq(Dummy(integer=True)/2, 0.5)`; this formerly failed because the lhs was rational and the rhs was not. Yet, SymPy recognizes that `Eq(S.Half, 0.5) -> True`. So, in fact, SymPy is treating 0.5 as a rational in this context. So the changes in this PR are aligned with that idea. The only alternative that I see to this is to make `Eq(S.Half, 0.5)` return False.

* `solve(x**.4 - 4)` still gives `[32.0]` because it takes action based on the exponent being Float instead of rational. So this is an example of something that did not "break" with this change. If, however, the code had made a decision based on the exponent being rational, it might have given a different result.

There are more than 600 uses of `is_integer` and about 200 uses of `is_rational` in the SymPy codebase.  Although no test currently fails there could be untested cases where a Float instead of a symbolic rational/integer will slip through into a codepath that was meant only for the symbolic (not Numeric) argument.

#### Other comments

Summary of things with assumptions:

```
Number (always is_rational)
* Integer is_Integer == is_Rational == is_integer == is_rational
* Rational (n/d) is_Rational == is_rational == True, collapses to Integer if d == 1
* Float (m/2**e) is_Integer == is_Rational == False, is_rational == True, is_integer if e==0, never collapses to Rational

Symbol is_integer per assumptions

NumberSymbol
* I is_imagary
* pi is_irrational
* Exp1 is_irrational
```

#### Release Notes

* core
  * Float is now treated like Rational in terms of assumptions, so `Eq(Dummy('x', integer=True), 1.0)` is not False
  * `e.is_integer` used to be true for Integer and symbolic integers; it is now true for Float if `e%1==0`.

### CODE REVIEW

**Comment by smichr** [comment]
I did not expect this to pass so easily. I suspect there are many corner cases lurking in the codebase.

File: sympy/printing/tests/test_c.py [hunk 1]
```diff
@@ -699,8 +699,8 @@ def check(expr, ref):
         check(gamma(x), 'tgamma{s}(x)')
         check(loggamma(x), 'lgamma{s}(x)')
 
-        check(ceiling(x + 2.), "ceil{s}(x + 2.0{S})")
-        check(floor(x + 2.), "floor{s}(x + 2.0{S})")
+        check(ceiling(x + 2.), "ceil{s}(x) + 2.0{S}")
+        check(floor(x + 2.), "floor{s}(x) + 2.0{S}")
```
Comment (asmeurer) [inline sympy/printing/tests/test_c.py:703]: This is potentially problematic, as I noted on the other PR, especially for the purposes of codegen. These two expressions aren't actually equal for floating point numbers. 

(but then again, SymPy does a lot of simplifications that aren't true in general for floating-point numbers)

File: sympy/core/numbers.py [hunk 2]
```diff
@@ -649,49 +660,122 @@ class Float(Number):
...
+    also be true for Floats:
+
+    >>> f1 = Float(1)
+    >>> f1.is_Rational, f1.is_rational, f1.is_integer, f1.is_odd
+    (False, True, True, True)
+    >>> r = Float(0.1)
+    >>> r.is_rational, r.is_integer, r.is_odd
+    (True, False, False)
+
+    == SymPy Float from Python float ==
```
Comment (asmeurer) [inline sympy/core/numbers.py:743]: Not sure how this sort of header will render in Sphinx.

File: sympy/printing/tests/test_c.py (hunk 1 above)
Comment (smichr) [reply sympy/printing/tests/test_c.py:703]: SymPy is returning a result that is `symbolically` correct, even if `x = Float(1e-16)`. It seems like that is in line with SymPy's mission. Guessing whether someone wants an expression to suffer truncation or not is not, so I think making these consistent is the way to go.  Please see #25872 for just this addition. And I think I prefer #25875 for the more robust and less controversial change to handle #23731.

File: sympy/core/numbers.py (hunk 2 above)
Comment (smichr) [reply sympy/core/numbers.py]: I redid these using `----` underlining

**Comment by smichr** [comment]
I am closing this in favor of #25875

**Pull request closed without merging**